waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -208,26 +208,58 @@
       (log/error e "error converting failed pod to waiter service instance" pod)
       (comment "Returning nil on failure."))))
 
+(defn- check-expired
+  "Returns true when the pod instance can be classified as expired.
+   An instance can be expired for the following reasons:
+   - it has restarted too many times (reached the restart-expiry-threshold threshold)
+   - the primary container (waiter-apps) has not transitioned to running state in container-running-grace-secs seconds."
+  [{:keys [container-running-grace-secs restart-expiry-threshold]}
+   instance-id restart-count primary-container-status pod-started-at]
+  (cond
+    (>= restart-count restart-expiry-threshold)
+    (do
+      (log/info "instance expired as it reached the restart threshold"
+                {:instance-id instance-id
+                 :restart-count restart-count})
+      true)
+    (and pod-started-at
+         (pos? container-running-grace-secs)
+         (not (contains? (get primary-container-status :state) :running))
```

Comment:
Nitpick: `(get primary-container-status :state)` → `(:state primary-container-status)`

Reply:
This instantaneous state check doesn't give us the right behavior. E.g., if a container starts, then crashes, then restarts, we'll mark the container expired because it's back in the `:waiting` state after `container-running-grace-secs` elapsed.

Possible workarounds:
1. Check that the restart count is zero.
2. Check that `:lastState` is not set (possibly less flaky than checking restart counts).
3. Keep extra state to see if the container was ever running.